File: test/apply.test.ts
import assert from "node:assert/strict";
import {
  mkdirSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  writeFileSync,
  existsSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { after, before, describe, it } from "node:test";

import { applyWorktreeInclude } from "../src/apply.js";
import { runGit } from "../src/git.js";

function initRepo(dir: string): void {
  runGit(["init", "-q"], { cwd: dir });
  runGit(["config", "user.email", "test@example.com"], { cwd: dir });
  runGit(["config", "user.name", "test"], { cwd: dir });
}

describe("applyWorktreeInclude", () => {
  let root: string;
  let main: string;
  let worktree: string;

  before(() => {
    root = mkdtempSync(join(tmpdir(), "herdr-wti-test-"));
    main = join(root, "main");
    worktree = join(root, "feature");
    mkdirSync(main);

    initRepo(main);
    writeFileSync(join(main, "README.md"), "hello\n");
    writeFileSync(join(main, ".gitignore"), ".env\n.env.local\nconfig/secrets.json\n");
    writeFileSync(join(main, ".env"), "SECRET=1\n");
    writeFileSync(join(main, ".env.local"), "LOCAL=1\n");
    writeFileSync(join(main, ".env.example"), "EXAMPLE=1\n");
    mkdirSync(join(main, "config"));
    writeFileSync(join(main, "config", "secrets.json"), '{"k":1}\n');
    writeFileSync(join(main, "config", "app.json"), '{"app":true}\n');
    writeFileSync(
      join(main, ".worktreeinclude"),
      [".env", ".env.local", "config/secrets.json", "!.env.example", ""].join(
        "\n",
      ),
    );

    runGit(["add", "README.md", ".gitignore", ".env.example", "config/app.json", ".worktreeinclude"], {
      cwd: main,
    });
    // .env.example is tracked; .env is ignored
    runGit(["add", "-f", ".env.example"], { cwd: main, check: false });
    runGit(
      ["commit", "-q", "-m", "init"],
      { cwd: main },
    );

    runGit(["branch", "feature"], { cwd: main });
    runGit(["worktree", "add", "-q", worktree, "feature"], { cwd: main });
  });

  after(() => {
    try {
      runGit(["worktree", "remove", "--force", worktree], {
        cwd: main,
        check: false,
      });
    } catch {
      // ignore
    }
    rmSync(root, { recursive: true, force: true });
  });

  it("copies only include∩ignored files", async () => {
    const result = await applyWorktreeInclude({
      sourceRoot: main,
      targetRoot: worktree,
    });

    assert.ok(result.includeFile);
    assert.deepEqual(new Set(result.copied), new Set([
      ".env",
      ".env.local",
      "config/secrets.json",
    ]));
    assert.equal(readFileSync(join(worktree, ".env"), "utf8"), "SECRET=1\n");
    assert.equal(
      readFileSync(join(worktree, "config/secrets.json"), "utf8"),
      '{"k":1}\n',
    );
    // tracked example should not be re-copied from include negation path
    assert.ok(existsSync(join(worktree, ".env.example")));
  });

  it("skips existing destinations and dry-run does not write", async () => {
    writeFileSync(join(worktree, ".env"), "ALREADY=1\n");
    const dry = await applyWorktreeInclude({
      sourceRoot: main,
      targetRoot: worktree,
      dryRun: true,
    });
    assert.ok(dry.skippedExisting.includes(".env"));
    assert.equal(readFileSync(join(worktree, ".env"), "utf8"), "ALREADY=1\n");

    const again = await applyWorktreeInclude({
      sourceRoot: main,
      targetRoot: worktree,
    });
    assert.ok(again.skippedExisting.includes(".env"));
    assert.equal(readFileSync(join(worktree, ".env"), "utf8"), "ALREADY=1\n");
  });

  it("no-ops when .worktreeinclude is missing", async () => {
    const bare = mkdtempSync(join(root, "bare-"));
    initRepo(bare);
    writeFileSync(join(bare, "a.txt"), "a\n");
    runGit(["add", "a.txt"], { cwd: bare });
    runGit(["commit", "-q", "-m", "init"], { cwd: bare });
    const wt = join(root, "bare-wt");
    runGit(["worktree", "add", "-q", wt, "-b", "b2", "HEAD"], { cwd: bare });

    const result = await applyWorktreeInclude({
      sourceRoot: bare,
      targetRoot: wt,
    });
    assert.equal(result.includeFile, null);
    assert.equal(result.copied.length, 0);
  });
});

File: tests/apply_test.rs
use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use herdr_worktreeinclude::apply::{apply_worktree_include, ApplyOptions};

fn run_git(args: &[&str], cwd: &Path) {
    let status = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .status()
        .expect("spawn git");
    assert!(status.success(), "git {args:?} failed");
}

fn run_git_ok(args: &[&str], cwd: &Path) {
    let _ = Command::new("git").args(args).current_dir(cwd).status();
}

fn init_repo(dir: &Path) {
    run_git(&["init", "-q"], dir);
    run_git(&["config", "user.email", "test@example.com"], dir);
    run_git(&["config", "user.name", "test"], dir);
}

fn write(path: &Path, contents: &str) {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).unwrap();
    }
    fs::write(path, contents).unwrap();
}

struct Fixture {
    _tmp: tempfile::TempDir,
    main: PathBuf,
    worktree: PathBuf,
}

fn setup() -> Fixture {
    let tmp = tempfile::Builder::new()
        .prefix("herdr-wti-test-")
        .tempdir()
        .unwrap();
    let main = tmp.path().join("main");
    let worktree = tmp.path().join("feature");
    fs::create_dir(&main).unwrap();

    init_repo(&main);
    write(&main.join("README.md"), "hello\n");
    write(
        &main.join(".gitignore"),
        ".env\n.env.local\nconfig/secrets.json\n",
    );
    write(&main.join(".env"), "SECRET=1\n");
    write(&main.join(".env.local"), "LOCAL=1\n");
    write(&main.join(".env.example"), "EXAMPLE=1\n");
    write(&main.join("config/secrets.json"), "{\"k\":1}\n");
    write(&main.join("config/app.json"), "{\"app\":true}\n");
    write(
        &main.join(".worktreeinclude"),
        ".env\n.env.local\nconfig/secrets.json\n!.env.example\n",
    );

    run_git(
        &[
            "add",
            "README.md",
            ".gitignore",
            ".env.example",
            "config/app.json",
            ".worktreeinclude",
        ],
        &main,
    );
    run_git_ok(&["add", "-f", ".env.example"], &main);
    run_git(&["commit", "-q", "-m", "init"], &main);
    run_git(&["branch", "feature"], &main);
    run_git(
        &[
            "worktree",
            "add",
            "-q",
            worktree.to_str().unwrap(),
            "feature",
        ],
        &main,
    );

    Fixture {
        _tmp: tmp,
        main,
        worktree,
    }
}

#[test]
fn copies_only_include_intersect_ignored_files() {
    let fix = setup();
    let result = apply_worktree_include(&ApplyOptions {
        source_root: fix.main.clone(),
        target_root: fix.worktree.clone(),
        dry_run: false,
    })
    .expect("apply");

    assert!(result.include_file.is_some());
    let copied: HashSet<_> = result.copied.into_iter().collect();
    let expected: HashSet<_> = [".env", ".env.local", "config/secrets.json"]
        .into_iter()
        .map(String::from)
        .collect();
    assert_eq!(copied, expected);
    assert_eq!(
        fs::read_to_string(fix.worktree.join(".env")).unwrap(),
        "SECRET=1\n"
    );
    assert_eq!(
        fs::read_to_string(fix.worktree.join("config/secrets.json")).unwrap(),
        "{\"k\":1}\n"
    );
    assert!(fix.worktree.join(".env.example").exists());
}

#[test]
fn skips_existing_and_dry_run_does_not_write() {
    let fix = setup();
    apply_worktree_include(&ApplyOptions {
        source_root: fix.main.clone(),
        target_root: fix.worktree.clone(),
        dry_run: false,
    })
    .unwrap();

    write(&fix.worktree.join(".env"), "ALREADY=1\n");
    let dry = apply_worktree_include(&ApplyOptions {
        source_root: fix.main.clone(),
        target_root: fix.worktree.clone(),
        dry_run: true,
    })
    .unwrap();
    assert!(dry.skipped_existing.iter().any(|p| p == ".env"));
    assert_eq!(
        fs::read_to_string(fix.worktree.join(".env")).unwrap(),
        "ALREADY=1\n"
    );

    let again = apply_worktree_include(&ApplyOptions {
        source_root: fix.main.clone(),
        target_root: fix.worktree.clone(),
        dry_run: false,
    })
    .unwrap();
    assert!(again.skipped_existing.iter().any(|p| p == ".env"));
    assert_eq!(
        fs::read_to_string(fix.worktree.join(".env")).unwrap(),
        "ALREADY=1\n"
    );
}

#[test]
fn noops_when_worktreeinclude_missing() {
    let tmp = tempfile::Builder::new()
        .prefix("herdr-wti-bare-")
        .tempdir()
        .unwrap();
    let bare = tmp.path().join("bare");
    let wt = tmp.path().join("bare-wt");
    fs::create_dir(&bare).unwrap();
    init_repo(&bare);
    write(&bare.join("a.txt"), "a\n");
    run_git(&["add", "a.txt"], &bare);
    run_git(&["commit", "-q", "-m", "init"], &bare);
    run_git(
        &[
            "worktree",
            "add",
            "-q",
            wt.to_str().unwrap(),
            "-b",
            "b2",
            "HEAD",
        ],
        &bare,
    );

    let result = apply_worktree_include(&ApplyOptions {
        source_root: bare,
        target_root: wt,
        dry_run: false,
    })
    .unwrap();
    assert!(result.include_file.is_none());
    assert!(result.copied.is_empty());
}
